ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/RecordFormatter.java

```diff
@@ -103,10 +103,25 @@ public RecordFormatter(
     this.valueDeserializers = requireNonNull(valueDeserializers, "valueDeserializers");
   }
 
-  public List<Supplier<String>> format(final Iterable<ConsumerRecord<Bytes, Bytes>> records) {
-    return StreamSupport.stream(records.spliterator(), false)
-        .map(this::delayedFormat)
-        .collect(Collectors.toList());
+  public List<String> format(final Iterable<ConsumerRecord<Bytes, Bytes>> records) {
+    final String activeKeyFormat = keyDeserializers.getPossibleFormats().get(0);
+    final String activeValueFormat = valueDeserializers.getPossibleFormats().get(0);
+
+    final List<String> formatted = formatRecords(records);
+
+    final boolean sameKeyFormatChanged = activeKeyFormat
+        .equals(keyDeserializers.getPossibleFormats().get(0));
+
+    final boolean sameValueFormatChanged = activeValueFormat
+        .equals(valueDeserializers.getPossibleFormats().get(0));
+
+    if (sameKeyFormatChanged && sameValueFormatChanged) {
+      return formatted;
+    }
+
+    // The active key/value format has been eliminated as a possibility while processing this batch.
+    // Reformat with the new active format:
+    return formatRecords(records);
```

Comment:
Why do call `formatRecords` here again when you already formatted them at line 110?

Reply:
As the comment says: because the active key or value format has changed during the processing of the batch.  

For example, if processing the second record in the batch eliminates the `KAFKA_BIGINT` key format, then the first row, which will have been formatted using `KAFKA_BIGINT` needs to be processed again, otherwise it will be printed out wrong.

This is the whole point of the PR!